django/db/models/constraints.py

```diff
@@ -119,3 +133,31 @@ def deconstruct(self):
         if self.condition:
             kwargs['condition'] = self.condition
         return path, args, kwargs
+
+
+DEFERRED = 'DEFERRED'
+IMMEDIATE = 'IMMEDIATE'
+
+
+@contextmanager
+def set_constraints(using=None, savepoint=True, **kwargs):
```

Comment:
This seems like a weird API. Is there really a need to make constraints immediate? What about only a `defer_constraints` API instead?

Reply:
I'm not sure if there is a use case. I did it this way as the most flexible option.

Reply:
I think we should avoid exposing a public API that we'll have to commit to maintaining for the upcoming releases then. The `connection` object [already has a few backend agnostics methods to disable constraints](https://github.com/django/django/blob/f110de5c04818b8f915dcf65da37a50c1424c6e6/django/db/backends/base/base.py#L443-L476) in the rare cases where this is required. I say we let the feature mature a bit and see how the API needs to be adjusted to meet users expectations. In all cases this function should probably delegate most of its SQL heavy lifting to the `connection` objects to allow third party backend overrides.

Reply:
I recently came across a use case for setting constraints immediate. It looks like this:

```python
class Author(Model):
    name = models.CharField(...)


class Book(Model):
    name = models.Charfield(...)
    author = models.ForeignKey(Author, ...)


@atomic
def create_book(name, author_id):
    try:
        return Book.objects.create(name=name, author_id=author_id)
    except IntegrityError:
        raise InvalidAuthor
```

Foreign keys in Django are deferred by default, so if the `author_id` is invalid, the `IntegrityError` is raised when the outer `atomic` decorator exists, not when the `Book.objects.create` method is called. I would need to set the author foreign key to immediate for this `IntegrityError` to be raised where I want to handle it.

```python-traceback
$ ./manage.py test
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
FE
======================================================================
ERROR: test_create (library.tests.TestCreateBook)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "django/test/testcases.py", line 274, in __call__
    self._post_teardown()
  File "django/test/testcases.py", line 1009, in _post_teardown
    self._fixture_teardown()
  File "django/test/testcases.py", line 1177, in _fixture_teardown
    connections[db_name].check_constraints()
  File "django/db/backends/sqlite3/base.py", line 318, in check_constraints
    bad_value, referenced_table_name, referenced_column_name
django.db.utils.IntegrityError: The row in table 'library_book' with primary key '1' has an invalid foreign key: library_book.author_id contains a value '1' that does not have a corresponding value in library_author.id.

======================================================================
FAIL: test_create (library.tests.TestCreateBook)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "django-sandbox/library/tests.py", line 9, in test_create
    create_book(name='The Hobbit', author_id=1)
AssertionError: InvalidAuthor not raised

----------------------------------------------------------------------
Ran 1 test in 0.002s

FAILED (failures=1, errors=1)
Destroying test database for alias 'default'...
```

Reply:
I've kept `set_constraints`, but simplified the implementation.

I realised that setting an already deferred constraint explicitly to deferred meant it became immediate on rollback, which seemed very confusing to users and a lot of hassle to avoid. Instead, the user is responsible for calling `set_constraints` a second time if they want that behaviour in that transaction. I expect this not to be necessary in the majority of use-cases.

Reply:
> I realised that setting an already deferred constraint explicitly to deferred meant it became immediate on rollback, which seemed very confusing to users and a lot of hassle to avoid. Instead, the user is responsible for calling set_constraints a second time if they want that behaviour in that transaction. I expect this not to be necessary in the majority of use-cases.

I think that's one of the reason this function needs a bit of discussion and should be handled in a different. It has a lot of implications with regards to transaction handling which would need to be defined (what use cases are we supporting?) and documented.

Reply:
Fair enough - I'll move it into the tests for now, since it's handy there.